File: 1022.sum-of-root-to-leaf-binary-numbers.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($value) { $this->val = $value; }
 * }
 */
class Solution {

    public $res = 0;

    /**
     * @param TreeNode $root
     * @return Integer
     */
    function sumRootToLeaf($root) {
        $this->sumRootToLeafHelp($root, 0);
        return $this->res;
    }
    
    function sumRootToLeafHelp($root, $prev) {
        if (empty($root)) {
            return;
        }

        $cur = $prev * 2 + $root->val;
        
        if (!empty($root->left)) {
            $this->sumRootToLeafHelp($root->left, $cur);
        }
        
        if (!empty($root->right)) {
            $this->sumRootToLeafHelp($root->right, $cur);
        }
        
        if (empty($root->left) && empty($root->right)) {
            $this->res += $cur;
        }
    }
}

File: 1022.sum-of-root-to-leaf-binary-numbers.cpp
// CreateTime: 2019-12-30 19:23:07
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    const int MOD = 1e9+7;
    int res = 0;

    int sumRootToLeaf(TreeNode* root) {
        dfs(root);
        return res;
    }
    
    void dfs(TreeNode *root, int v = 0) {
        if (root == NULL) {
            return;
        }

        v = v * 2 + root->val;
        
        if (root->left == NULL && root->right == NULL) {
            res = ((long long)res + (long long)v) % MOD;
            return;
        }

        dfs(root->left, v);
        dfs(root->right, v);
    }
};
